firedrake/utility_meshes.py

```diff
@@ -568,3 +569,254 @@ def UnitIcosahedralSphereMesh(refinement_level=0, reorder=None):
+    local_faces[:, 1] = vals
+    vals = vertex_numbers[:-1, 1:].reshape(((Nx-1)**2,))
+    local_faces[:, 3] = vals
+    vals = vertex_numbers[1:, 1:].reshape(((Nx-1)**2,))
+    local_faces[:, 2] = vals
+
+    p0 = panel_numbering[0, :, :].reshape((Nx**2,))
+    p1 = panel_numbering[1, :, :].reshape((Nx**2,))
+    p2 = panel_numbering[2, :, :].reshape((Nx**2,))
+    p3 = panel_numbering[3, :, :].reshape((Nx**2,))
+    p4 = panel_numbering[4, :, :].reshape((Nx**2,))
+    p5 = panel_numbering[5, :, :].reshape((Nx**2,))
+    f0 = p0[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+    f1 = p1[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+    f2 = p2[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+    f3 = p3[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+    f4 = p4[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+    f5 = p5[local_faces.reshape((Nx-1)**2*4,)].reshape((Nx-1)**2, 4)
+
+    faces = np.concatenate((f0, f1, f2, f3, f4, f5))
+    return vertices, faces
+
+
+def CubedSphereMesh(radius, refinement_level=0, reorder=None,
+                    use_dmplex_refinement=False):
+    """Generate an cubed approximation to the surface of the
+    sphere.
+
+    :arg radius: The radius of the sphere to approximate.
+    :kwarg refinement_level: optional number of refinements (0 is a cube).
+    :kwarg reorder: (optional), should the mesh be reordered?
+    :kwarg use_dmplex_refinement: (optional), use dmplex to apply
+    the refinement.
+    """
+    if(use_dmplex_refinement):
+        # vertices of a cube with an edge length of 2
+        vertices = np.array([[-1., -1., -1.],
+                             [1., -1., -1.],
+                             [-1., 1., -1.],
+                             [1., 1., -1.],
+                             [-1., -1., 1.],
+                             [1., -1., 1.],
+                             [-1., 1., 1.],
+                             [1., 1., 1.]])
+        # faces of the base cube
+        # bottom face viewed from above
+        # 2 3
+        # 0 1
+        # top face viewed from above
+        # 6 7
+        # 4 5
+        faces = np.array([[0, 1, 3, 2],  # bottom
+                          [4, 5, 7, 6],  # top
+                          [0, 1, 5, 4],
+                          [2, 3, 7, 6],
+                          [0, 2, 6, 4],
+                          [1, 3, 7, 5]], dtype=np.int32)
+
+        plex = _from_cell_list(2, faces, vertices)
+        plex.setRefinementUniform(True)
+        for i in range(refinement_level):
+            plex = plex.refine()
+
+        #rescale points to the sphere
+        #this is not the same as the gnonomic transformation
+        vStart, vEnd = plex.getDepthStratum(0)
+        nvertices = vEnd - vStart
+        coords = plex.getCoordinatesLocal().array.reshape(nvertices, 3)
+        scale = (radius / np.linalg.norm(coords, axis=1)).reshape(-1, 1)
+        coords *= scale
```

Comment:
On 06/01/15 12:06, Doru Rathgeber wrote:
> In firedrake/utility_meshes.py
>> +
>> +        #rescale points to the sphere
>> +        #this is not the same as the gnonomic transformation
>> +        vStart, vEnd = plex.getDepthStratum(0)
>> +        nvertices = vEnd - vStart
>> +        coords = plex.getCoordinatesLocal().array.reshape(nvertices, 3)
>> +        scale = (radius / np.linalg.norm(coords, axis=1)).reshape(-1, 1)
>> +        coords *= scale
>
> Did this get fixed? If not, what needs to be done here?

It causes PETSc to segfault, I don't know why. I think this type of 
thing hasn't been done in PETSc before.

--cjc